cpp/ql/src/semmle/code/cpp/controlflow/StackVariableReachability.qll

```diff
@@ -3,7 +3,194 @@
  * reachability involving stack variables.
  */
 
-import cpp
+private import semmle.code.cpp.controlflow.Guards
+private import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+private newtype TCondition =
+  TTrue() or
+  TLogicalCondition(LogicalGuardCondition guard, boolean testIsTrue) { testIsTrue = [false, true] }
+
+/** A `GuardCondition` which appear in a control-flow path to a sink. */
+abstract private class LogicalGuardCondition extends GuardCondition {
+  LogicalGuardCondition() {
+    // Either the `GuardCondition` is part of the path from a source to a sink
+    revBbSuccessorEntryReaches0(_, this.getBasicBlock(), _, _, _)
+    or
+    // or it controls the basic block that contains the source node.
+    this.controls(any(BasicBlock bb | fwdBbEntryReachesLocally(bb, _, _, _)), _)
+  }
+
+  /**
+   * Holds if the truth of this logical expression having value `wholeIsTrue`
+   * implies that the truth of the child expression `part` has truth value `partIsTrue`.
+   */
+  abstract predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  );
+}
+
+private class BinaryLogicalGuardCondition extends LogicalGuardCondition, BinaryLogicalOperation {
+  override predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  ) {
+    this.impliesValue(e, testIsTrue, condIsTrue)
+  }
+}
+
+private class VariableGuardCondition extends LogicalGuardCondition, VariableAccess {
+  override predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  ) {
+    this = e and
+    (
+      testIsTrue = true and condIsTrue = true
+      or
+      testIsTrue = false and condIsTrue = false
+    )
+  }
+}
+
+private class NotGuardCondition extends LogicalGuardCondition, NotExpr {
+  override predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  ) {
+    e = this.getOperand() and
+    (
+      testIsTrue = true and
+      condIsTrue = false
+      or
+      testIsTrue = false and
+      condIsTrue = true
+    )
+  }
+}
+
+/** A condition that either the trivial `True` condition, or a complex `LogicalCondition`. */
```

Comment:
So there are (apart from the trivial `True` condition) a combination of a `LogicalCondition` and a truth value we suppose it takes.

This class feels like it might at some point belong in a separate `.qll` file, closer to the guards library than to stack variable reachability.

Why do we not need / have the trivial `False` condition?

Reply:
Thanks for this comment. While trying to explain the role of the `True` condition I realized that it wasn't actually needed anymore. So I've removed it in c8c77396fac4a70d66b82f8b64cb1ddd5484008a. My original intention with `True` was that it would make `impliesCondition` always have a result (since this was needed in a previous iteration of the code). But upon looking at the state of the code now, I realized that this wasn't actually needed anymore.

The reason for not adding a `False` condition was that there was no equivalent requirement that `refutesCondition` always had a result (like the role of `True` was to ensure that `impliesCondition` always had a result).

I agree that it would be useful to have `Condition` in a separate `.qll` file closer to the guards library. The problem is that a `LogicalCondition` depends on `LogicalGuardCondition`, and (for performance) the corresponding IPA type `TLogicalGuardCondition` is restricted to be only conditions that appear in the control-flow path computed by the `revBbSuccessorEntryReaches0` predicate.